crates/ethereum/evm/src/factory.rs

```diff
@@ -0,0 +1,86 @@
+use reth_evm::{
+    eth::EthEvmContext, Database, EthEvm, EvmEnv, EvmFactory, MaybeCachedPrecompileProvider,
+};
+use revm::{
+    context::{
+        result::{EVMError, HaltReason},
+        BlockEnv, CfgEnv, TxEnv,
+    },
+    handler::EthPrecompiles,
+    inspector::NoOpInspector,
+    primitives::hardfork::SpecId,
+    Context, Inspector, MainBuilder, MainContext,
+};
+
+/// Factory producing [`MaybeCachedPrecompileEthEvmFactory`].
+#[derive(Debug, Default, Clone)]
+#[non_exhaustive]
+pub struct MaybeCachedPrecompileEthEvmFactory {
```

Comment:
I think we can even add this logic to the regular ethevmfactory if we make it generic over a provider that is Clone?

like 
```
EthEvmFactory<P= EthPrecompiles> {
  precompiles: P
}

https://github.com/alloy-rs/evm/blob/05f7c16a744af7eb63a068d6b3a5a7e9a2504996/crates/evm/src/eth/mod.rs#L208-L210

?